### ISSUE TITLE
Fatal Exception: java.lang.RuntimeException: android.os.DeadObjectException

### ISSUE THREAD

**Issue by eveliotc** [firebase/firebase-jobdispatcher-android#26]
Received this through crash reporting:

```
Fatal Exception: java.lang.RuntimeException: android.os.DeadObjectException
       at com.firebase.jobdispatcher.GooglePlayJobCallback.jobFinished(GooglePlayJobCallback.java:39)
       at com.firebase.jobdispatcher.GooglePlayReceiver.onJobFinished(GooglePlayReceiver.java:162)
       at com.firebase.jobdispatcher.ExternalReceiver.onJobFinishedMessage(ExternalReceiver.java:69)
       at com.firebase.jobdispatcher.ExternalReceiver.access$200(ExternalReceiver.java:41)
       at com.firebase.jobdispatcher.ExternalReceiver$ResponseHandler.handleMessage(ExternalReceiver.java:115)
       at android.os.Handler.dispatchMessage(Handler.java:102)
       at android.os.Looper.loop(Looper.java:158)
       at android.app.ActivityThread.main(ActivityThread.java:7224)
       at java.lang.reflect.Method.invoke(Method.java)
       at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:1230)
       at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1120)
Caused by android.os.DeadObjectException
       at android.os.BinderProxy.transactNative(Binder.java)
       at android.os.BinderProxy.transact(Binder.java:503)
       at com.google.android.gms.gcm.INetworkTaskCallback$Stub$Proxy.taskFinished(INetworkTaskCallback.java:85)
       at com.firebase.jobdispatcher.GooglePlayJobCallback.jobFinished(GooglePlayJobCallback.java:37)
       at com.firebase.jobdispatcher.GooglePlayReceiver.onJobFinished(GooglePlayReceiver.java:162)
       at com.firebase.jobdispatcher.ExternalReceiver.onJobFinishedMessage(ExternalReceiver.java:69)
       at com.firebase.jobdispatcher.ExternalReceiver.access$200(ExternalReceiver.java:41)
       at com.firebase.jobdispatcher.ExternalReceiver$ResponseHandler.handleMessage(ExternalReceiver.java:115)
       at android.os.Handler.dispatchMessage(Handler.java:102)
       at android.os.Looper.loop(Looper.java:158)
       at android.app.ActivityThread.main(ActivityThread.java:7224)
       at java.lang.reflect.Method.invoke(Method.java)
       at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:1230)
       at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1120)
```

My current theory is that this is caused because `jobFinished` is being called after `onStopJob` so will add measures to prevent this will report if this continues to happen.

**Comment by eveliotc** [comment]
This continues to happen, ensuring `jobFinished` is not called after `onStopJob` didn't make a difference, happening on the following devices at the moment

- SM-G930T (Android 6.0.1)
- SM-G530T1 (Android 5.1.1)

**Comment by YohayBar** [comment]
same here:
```Fatal Exception: java.lang.RuntimeException: android.os.DeadObjectException
       at com.firebase.jobdispatcher.GooglePlayJobCallback.jobFinished(GooglePlayJobCallback.java:39)
       at com.firebase.jobdispatcher.GooglePlayReceiver.onJobFinished(GooglePlayReceiver.java:162)
       at com.firebase.jobdispatcher.ExternalReceiver.onJobFinishedMessage(ExternalReceiver.java:69)
       at com.firebase.jobdispatcher.ExternalReceiver.access$200(ExternalReceiver.java:41)
       at com.firebase.jobdispatcher.ExternalReceiver$ResponseHandler.handleMessage(ExternalReceiver.java:115)
       at android.os.Handler.dispatchMessage(Handler.java:102)
       at android.os.Looper.loop(Looper.java:136)
       at android.app.ActivityThread.main(ActivityThread.java:5584)
       at java.lang.reflect.Method.invokeNative(Method.java)
       at java.lang.reflect.Method.invoke(Method.java:515)
       at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:1268)
       at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1084)
       at dalvik.system.NativeStart.main(NativeStart.java)
Caused by android.os.DeadObjectException
       at android.os.BinderProxy.transact(Binder.java)
       at com.google.android.gms.gcm.INetworkTaskCallback$Stub$Proxy.taskFinished(INetworkTaskCallback.java:85)
       at com.firebase.jobdispatcher.GooglePlayJobCallback.jobFinished(GooglePlayJobCallback.java:37)
       at com.firebase.jobdispatcher.GooglePlayReceiver.onJobFinished(GooglePlayReceiver.java:162)
       at com.firebase.jobdispatcher.ExternalReceiver.onJobFinishedMessage(ExternalReceiver.java:69)
       at com.firebase.jobdispatcher.ExternalReceiver.access$200(ExternalReceiver.java:41)
       at com.firebase.jobdispatcher.ExternalReceiver$ResponseHandler.handleMessage(ExternalReceiver.java:115)
       at android.os.Handler.dispatchMessage(Handler.java:102)
       at android.os.Looper.loop(Looper.java:136)
       at android.app.ActivityThread.main(ActivityThread.java:5584)
       at java.lang.reflect.Method.invokeNative(Method.java)
       at java.lang.reflect.Method.invoke(Method.java:515)
       at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:1268)
       at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1084)
       at dalvik.system.NativeStart.main(NativeStart.java)```

**Comment by maksim88** [comment]
Got the same one:
```
Fatal Exception: java.lang.RuntimeException: android.os.DeadObjectException
       at com.firebase.jobdispatcher.GooglePlayJobCallback.jobFinished(GooglePlayJobCallback.java:39)
       at com.firebase.jobdispatcher.GooglePlayReceiver.onJobFinished(GooglePlayReceiver.java:162)
       at com.firebase.jobdispatcher.ExternalReceiver.onJobFinishedMessage(ExternalReceiver.java:69)
       at com.firebase.jobdispatcher.ExternalReceiver.access$200(ExternalReceiver.java:41)
       at com.firebase.jobdispatcher.ExternalReceiver$ResponseHandler.handleMessage(ExternalReceiver.java:115)
       at android.os.Handler.dispatchMessage(Handler.java:99)
       at android.os.Looper.loop(Looper.java:176)
       at android.app.ActivityThread.main(ActivityThread.java:5419)
       at java.lang.reflect.Method.invokeNative(Method.java)
       at java.lang.reflect.Method.invoke(Method.java:525)
       at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:1046)
       at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:862)
       at dalvik.system.NativeStart.main(NativeStart.java)
Caused by android.os.DeadObjectException
       at android.os.BinderProxy.transact(Binder.java)
       at com.google.android.gms.gcm.INetworkTaskCallback$Stub$Proxy.taskFinished(INetworkTaskCallback.java:85)
       at com.firebase.jobdispatcher.GooglePlayJobCallback.jobFinished(GooglePlayJobCallback.java:37)
       at com.firebase.jobdispatcher.GooglePlayReceiver.onJobFinished(GooglePlayReceiver.java:162)
       at com.firebase.jobdispatcher.ExternalReceiver.onJobFinishedMessage(ExternalReceiver.java:69)
       at com.firebase.jobdispatcher.ExternalReceiver.access$200(ExternalReceiver.java:41)
       at com.firebase.jobdispatcher.ExternalReceiver$ResponseHandler.handleMessage(ExternalReceiver.java:115)
       at android.os.Handler.dispatchMessage(Handler.java:99)
       at android.os.Looper.loop(Looper.java:176)
       at android.app.ActivityThread.main(ActivityThread.java:5419)
       at java.lang.reflect.Method.invokeNative(Method.java)
       at java.lang.reflect.Method.invoke(Method.java:525)
       at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:1046)
       at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:862)
       at dalvik.system.NativeStart.main(NativeStart.java)
```

Affected devices: 
`Samsung GT-I9300 (Android 4.3)`
`Asus K00F (Android 4.2.2)`

**Comment by yahor-makouski-epam** [comment]
+1
Experiencing the same problem

**Comment by eveliotc** [comment]
This is averaging ~100 daily crashes, during Google Play Services outage last Friday we saw a spike of 900+ crashes.

**Comment by ciarand** [comment]
Are you all running against the version that includes https://github.com/firebase/firebase-jobdispatcher-android/commit/43ea5324856e08d500fd2991783acd338144029c?

**Comment by yahor-makouski-epam** [comment]
Yes, the crash still persists in the latest version with the fix https://github.com/firebase/firebase-jobdispatcher-android/commit/43ea5324856e08d500fd2991783acd338144029c.

Here are the steps to reproduce the crash:
1. start long running job
2. kill com.google.android.gms/.gcm.GcmService. I have a rooted device so I can do that via adb shell kill PID
3. when long running job is finished we invoke jobFinished() method

DeadObjectException occurs in GooglePlayJobCallback#jobFinished()
I think we should not throw RuntimeException when any RemoteException occurs.

And also when I kill GcmService the service is restarted automatically but no jobs are rescheduled. But this is something to be fixed on GCM library side.

**Comment by outlying** [comment]
Same stacktrace for me, my job scheduling code can be found [here](https://github.com/SmokSmog/smoksmog-android/blob/4e8495f3550c03414b9d11ecdf789a8e1614f71a/app/src/main/java/com/antyzero/smoksmog/SmokSmogApplication.java#L70) and it will trigger actions in this [service](https://github.com/SmokSmog/smoksmog-android/blob/4e8495f3550c03414b9d11ecdf789a8e1614f71a/app/src/main/kotlin/com/antyzero/smoksmog/job/SmokSmogJobService.kt) (tl;dr it force widgets to update more frequently)

**Comment by ciarand** [comment]
We're reviewing a fix for this and hope to have it out this week. We'll tag a new version when that happens.

**Comment by kusan-thana** [comment]
Is there a fix yet?

Because otherwise I'll have to go look at official API....

Thank you in advance,

**Comment by yahor-makouski-epam** [comment]
@ciarand could you please release a new version with the fix?
It's very important for us.

**Comment by ciarand** [comment]
Yes, I'm just chasing a couple of other PRs that would be good to get into this release. Regardless of what else happens we'll have a new version within 8 hours or so.

**Comment by ciarand** [comment]
Cut v0.5.1, which should be available on JCenter now.

**Comment by Ansh1234** [comment]
Why is this issue closed when it is still not fixed? Also, there is no v0.5.1 pushed out yet. Getting close to 100 crashes because of this. Please help fix this soon.

**Comment by bobroalexandr** [comment]
Still facing that issues even using v0.5.2

`Fatal Exception: java.lang.RuntimeException: android.os.DeadObjectException
       at com.firebase.jobdispatcher.GooglePlayJobCallback.jobFinished(GooglePlayJobCallback.java:38)
       at com.firebase.jobdispatcher.GooglePlayReceiver.onJobFinished(GooglePlayReceiver.java:161)
       at com.firebase.jobdispatcher.ExternalReceiver.onJobFinishedMessage(ExternalReceiver.java:72)
       at com.firebase.jobdispatcher.ExternalReceiver.access$200(ExternalReceiver.java:40)
       at com.firebase.jobdispatcher.ExternalReceiver$ResponseHandler.handleMessage(ExternalReceiver.java:118)
       at android.os.Handler.dispatchMessage(Handler.java:102)
       at android.os.Looper.loop(Looper.java:145)
       at android.app.ActivityThread.main(ActivityThread.java:6145)
       at java.lang.reflect.Method.invoke(Method.java)
       at java.lang.reflect.Method.invoke(Method.java:372)
       at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:1399)
       at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1194)
Caused by android.os.DeadObjectException
       at android.os.BinderProxy.transactNative(Binder.java)
       at android.os.BinderProxy.transact(Binder.java:496)
       at com.google.android.gms.gcm.INetworkTaskCallback$Stub$Proxy.taskFinished(INetworkTaskCallback.java:85)
       at com.firebase.jobdispatcher.GooglePlayJobCallback.jobFinished(GooglePlayJobCallback.java:36)
       at com.firebase.jobdispatcher.GooglePlayReceiver.onJobFinished(GooglePlayReceiver.java:161)
       at com.firebase.jobdispatcher.ExternalReceiver.onJobFinishedMessage(ExternalReceiver.java:72)
       at com.firebase.jobdispatcher.ExternalReceiver.access$200(ExternalReceiver.java:40)
       at com.firebase.jobdispatcher.ExternalReceiver$ResponseHandler.handleMessage(ExternalReceiver.java:118)
       at android.os.Handler.dispatchMessage(Handler.java:102)
       at android.os.Looper.loop(Looper.java:145)
       at android.app.ActivityThread.main(ActivityThread.java:6145)
       at java.lang.reflect.Method.invoke(Method.java)
       at java.lang.reflect.Method.invoke(Method.java:372)
       at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:1399)
       at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1194)`
